File: appwrite_storage.py
"""
Appwrite Storage Service for Trading Post
Provides image upload, optimization, and CDN functionality
"""

from appwrite.exception import AppwriteException
from appwrite.id import ID
from appwrite.input_file import InputFile
from appwrite_config import appwrite_config, BUCKETS
import logging
from typing import Optional, Dict, Any, List
import os
import io
from PIL import Image
import filetype

logger = logging.getLogger(__name__)


class AppwriteStorageService:
    """Handle storage operations with Appwrite"""

    def __init__(self):
        self.client = appwrite_config.client
        self.storage = appwrite_config.storage

    def _optimize_image(self, image_data: bytes, max_size: tuple = (1200, 1200), quality: int = 85) -> bytes:
        """Optimize image for web delivery"""
        try:
            # Open image
            image = Image.open(io.BytesIO(image_data))

            # Convert to RGB if necessary
            if image.mode in ("RGBA", "P"):
                image = image.convert("RGB")

            # Resize if larger than max_size
            if image.size[0] > max_size[0] or image.size[1] > max_size[1]:
                image.thumbnail(max_size, Image.Resampling.LANCZOS)

            # Save optimized image
            output = io.BytesIO()
            image.save(output, format="JPEG", quality=quality, optimize=True)
            output.seek(0)

            return output.getvalue()

        except Exception as e:
            logger.warning(f"Image optimization failed: {e}, using original")
            return image_data

    def _create_thumbnail(self, image_data: bytes, size: tuple = (300, 300)) -> bytes:
        """Create thumbnail from image"""
        try:
            image = Image.open(io.BytesIO(image_data))

            # Convert to RGB if necessary
            if image.mode in ("RGBA", "P"):
                image = image.convert("RGB")

            # Create thumbnail
            image.thumbnail(size, Image.Resampling.LANCZOS)

            # Save thumbnail
            output = io.BytesIO()
            image.save(output, format="JPEG", quality=85, optimize=True)
            output.seek(0)

            return output.getvalue()

        except Exception as e:
            logger.warning(f"Thumbnail creation failed: {e}")
            return None

    async def upload_item_image(
        self, image_data: bytes, filename: str, user_id: str, listing_id: str = None
    ) -> Dict[str, Any]:
        """Upload and optimize item image"""
        try:
            # Validate file type
            file_type = filetype.guess(image_data)
            if not file_type or not file_type.mime.startswith("image/"):
                return {"error": "Invalid image file type", "success": False}

            # Optimize main image
            optimized_data = self._optimize_image(image_data)

            # Create thumbnail
            thumbnail_data = self._create_thumbnail(image_data)

            # Generate unique file ID
            file_id = ID.unique()

            # Upload main image
            main_file = await self.storage.create_file(
                bucket_id=BUCKETS["item_images"], file_id=file_id, file=InputFile.from_bytes(optimized_data, filename)
            )

            # Upload thumbnail if created successfully
            thumbnail_file = None
            if thumbnail_data:
                thumbnail_id = f"{file_id}_thumb"
                thumbnail_filename = f"thumb_{filename}"

                try:
                    thumbnail_file = await self.storage.create_file(
                        bucket_id=BUCKETS["item_images"],
                        file_id=thumbnail_id,
                        file=InputFile.from_bytes(thumbnail_data, thumbnail_filename),
                    )
                except Exception as e:
                    logger.warning(f"Thumbnail upload failed: {e}")

            # Get file URLs
            main_url = f"{appwrite_config.endpoint}/storage/buckets/{BUCKETS['item_images']}/files/{file_id}/view"
            thumbnail_url = (
                f"{appwrite_config.endpoint}/storage/buckets/{BUCKETS['item_images']}/files/{thumbnail_id}/view"
                if thumbnail_file
                else main_url
            )

            result = {
                "file_id": file_id,
                "url": main_url,
                "thumbnail_url": thumbnail_url,
                "filename": filename,
                "size": len(optimized_data),
                "original_size": len(image_data),
                "mime_type": file_type.mime,
                "success": True,
            }

            logger.info(f"Item image uploaded: {file_id}")
            return result

        except AppwriteException as e:
            logger.error(f"Failed to upload item image: {e}")
            return {"error": str(e), "code": e.code, "success": False}
        except Exception as e:
            logger.error(f"Image processing error: {e}")
            return {"error": str(e), "success": False}

    async def upload_profile_image(self, image_data: bytes, filename: str, user_id: str) -> Dict[str, Any]:
        """Upload and optimize profile image"""
        try:
            # Validate file type
            file_type = filetype.guess(image_data)
            if not file_type or not file_type.mime.startswith("image/"):
                return {"error": "Invalid image file type", "success": False}

            # Optimize image (smaller size for profiles)
            optimized_data = self._optimize_image(image_data, max_size=(800, 800), quality=90)

            # Create profile thumbnail
            thumbnail_data = self._create_thumbnail(image_data, size=(150, 150))

            # Use user_id as file_id for profiles (allows easy replacement)
            file_id = f"profile_{user_id}"

            # Delete existing profile image if it exists
            try:
                await self.storage.delete_file(bucket_id=BUCKETS["profile_images"], file_id=file_id)
            except AppwriteException:
                # File doesn't exist, that's fine
                pass

            # Upload main profile image
            main_file = await self.storage.create_file(
                bucket_id=BUCKETS["profile_images"],
                file_id=file_id,
                file=InputFile.from_bytes(optimized_data, filename),
            )

            # Upload thumbnail
            thumbnail_file = None
            if thumbnail_data:
                thumbnail_id = f"{file_id}_thumb"

                try:
                    # Delete existing thumbnail
                    await self.storage.delete_file(bucket_id=BUCKETS["profile_images"], file_id=thumbnail_id)
                except AppwriteException:
                    pass

                thumbnail_file = await self.storage.create_file(
                    bucket_id=BUCKETS["profile_images"],
                    file_id=thumbnail_id,
                    file=InputFile.from_bytes(thumbnail_data, f"thumb_{filename}"),
                )

            # Get file URLs
            main_url = f"{appwrite_config.endpoint}/storage/buckets/{BUCKETS['profile_images']}/files/{file_id}/view"
            thumbnail_url = (
                f"{appwrite_config.endpoint}/storage/buckets/{BUCKETS['profile_images']}/files/{thumbnail_id}/view"
                if thumbnail_file
                else main_url
            )

            result = {
                "file_id": file_id,
                "url": main_url,
                "thumbnail_url": thumbnail_url,
                "filename": filename,
                "size": len(optimized_data),
                "original_size": len(image_data),
                "mime_type": file_type.mime,
                "success": True,
            }

            logger.info(f"Profile image uploaded: {user_id}")
            return result

        except AppwriteException as e:
            logger.error(f"Failed to upload profile image: {e}")
            return {"error": str(e), "code": e.code, "success": False}
        except Exception as e:
            logger.error(f"Profile image processing error: {e}")
            return {"error": str(e), "success": False}

    async def delete_file(self, bucket_id: str, file_id: str) -> Dict[str, Any]:
        """Delete a file from storage"""
        try:
            await self.storage.delete_file(bucket_id=bucket_id, file_id=file_id)

            logger.info(f"File deleted: {file_id}")
            return {"success": True}

        except AppwriteException as e:
            logger.error(f"Failed to delete file {file_id}: {e}")
            return {"error": str(e), "code": e.code, "success": False}

    async def delete_item_images(self, file_ids: List[str]) -> Dict[str, Any]:
        """Delete multiple item images"""
        deleted_count = 0
        errors = []

        for file_id in file_ids:
            # Delete main image
            result = await self.delete_file(BUCKETS["item_images"], file_id)
            if result.get("success"):
                deleted_count += 1
            else:
                errors.append(f"Failed to delete {file_id}: {result.get('error')}")

            # Delete thumbnail
            thumbnail_id = f"{file_id}_thumb"
            await self.delete_file(BUCKETS["item_images"], thumbnail_id)

        return {"deleted_count": deleted_count, "errors": errors, "success": deleted_count > 0}

    async def get_file_info(self, bucket_id: str, file_id: str) -> Dict[str, Any]:
        """Get file information"""
        try:
            file_info = await self.storage.get_file(bucket_id=bucket_id, file_id=file_id)

            return {"file_info": file_info, "success": True}

        except AppwriteException as e:
            logger.error(f"Failed to get file info {file_id}: {e}")
            return {"error": str(e), "code": e.code, "success": False}

    async def get_file_preview(
        self, bucket_id: str, file_id: str, width: int = None, height: int = None, quality: int = None
    ) -> str:
        """Get optimized file preview URL"""
        try:
            preview_url = f"{appwrite_config.endpoint}/storage/buckets/{bucket_id}/files/{file_id}/preview"

            params = []
            if width:
                params.append(f"width={width}")
            if height:
                params.append(f"height={height}")
            if quality:
                params.append(f"quality={quality}")

            if params:
                preview_url += "?" + "&".join(params)

            return preview_url

        except Exception as e:
            logger.error(f"Failed to generate preview URL: {e}")
            return ""

    async def get_file_download_url(self, bucket_id: str, file_id: str) -> str:
        """Get file download URL"""
        return f"{appwrite_config.endpoint}/storage/buckets/{bucket_id}/files/{file_id}/download"

    async def list_user_files(self, user_id: str, bucket_id: str = None) -> Dict[str, Any]:
        """List files for a specific user"""
        try:
            buckets_to_check = [bucket_id] if bucket_id else list(BUCKETS.values())
            all_files = []

            for bucket in buckets_to_check:
                try:
                    files = await self.storage.list_files(bucket_id=bucket)

                    # Filter files that belong to this user (based on naming convention)
                    user_files = []
                    for file in files["files"]:
                        # Check if file belongs to user based on metadata or naming
                        if (
                            file["name"].startswith(f"user_{user_id}_")
                            or file["$id"].startswith(f"profile_{user_id}")
                            or user_id in file.get("metadata", {}).get("user_id", "")
                        ):
                            user_files.append(
                                {
                                    **file,
                                    "bucket_id": bucket,
                                    "url": f"{appwrite_config.endpoint}/storage/buckets/{bucket}/files/{file['$id']}/view",
                                }
                            )

                    all_files.extend(user_files)

                except AppwriteException as e:
                    logger.warning(f"Failed to list files in bucket {bucket}: {e}")

            return {"files": all_files, "success": True}

        except Exception as e:
            logger.error(f"Failed to list user files: {e}")
            return {"error": str(e), "success": False}

    async def get_storage_usage(self, user_id: str) -> Dict[str, Any]:
        """Get storage usage stats for user"""
        try:
            user_files_result = await self.list_user_files(user_id)

            if not user_files_result.get("success"):
                return user_files_result

            files = user_files_result["files"]

            usage_stats = {
                "total_files": len(files),
                "total_size_bytes": sum(file.get("sizeOriginal", 0) for file in files),
                "profile_images": len([f for f in files if "profile_images" in f.get("bucket_id", "")]),
                "item_images": len([f for f in files if "item_images" in f.get("bucket_id", "")]),
                "total_size_mb": sum(file.get("sizeOriginal", 0) for file in files) / 1024 / 1024,
            }

            return {"usage": usage_stats, "success": True}

        except Exception as e:
            logger.error(f"Failed to get storage usage: {e}")
            return {"error": str(e), "success": False}


# Global instance
appwrite_storage = AppwriteStorageService()

File: trading-app-frontend/src/services/appwriteStorage.js
/**
 * Appwrite Storage Service
 * Handles file uploads, image optimization, and CDN delivery
 */

import { storage, BUCKETS, handleAppwriteError, ID, getFilePreview, getFileView } from '../lib/appwrite';

class AppwriteStorageService {
  // Image Upload for Listings
  async uploadItemImage(file, userId, listingId = null) {
    try {
      console.log('📸 Uploading item image...');
      
      // Validate file object first
      if (!file) {
        throw new Error('No file provided for upload');
      }
      
      if (!(file instanceof File)) {
        throw new Error('Invalid file object - must be a File instance');
      }
      
      if (file.size === 0) {
        throw new Error('File is empty - no content to upload');
      }
      
      // Validate file type
      const allowedTypes = ['image/jpeg', 'image/jpg', 'image/png', 'image/webp'];
      if (!allowedTypes.includes(file.type)) {
        throw new Error('Invalid file type. Please upload JPEG, PNG, or WebP images.');
      }
      
      // Validate file size (max 10MB)
      const maxSize = 10 * 1024 * 1024; // 10MB
      if (file.size > maxSize) {
        throw new Error('File too large. Please upload images smaller than 10MB.');
      }
      
      // Generate unique file ID
      const fileId = ID.unique();
      
      // Upload file
      const uploadedFile = await storage.createFile(
        BUCKETS.itemImages,
        fileId,
        file
      );
      
      // Get file URLs
      const fileUrl = getFileView(BUCKETS.itemImages, fileId);
      const thumbnailUrl = getFilePreview(BUCKETS.itemImages, fileId, 300, 300, 85);
      const previewUrl = getFilePreview(BUCKETS.itemImages, fileId, 800, 600, 90);
      
      console.log('✅ Item image uploaded:', fileId);
      
      return {
        file: uploadedFile,
        fileId: fileId,
        url: fileUrl,
        thumbnailUrl: thumbnailUrl,
        previewUrl: previewUrl,
        filename: file.name,
        size: file.size,
        mimeType: file.type,
        success: true
      };
      
    } catch (error) {
      console.error('❌ Upload item image failed:', error);
      return {
        success: false,
        error: handleAppwriteError(error)
      };
    }
  }

  // Profile Image Upload
  async uploadProfileImage(file, userId) {
    try {
      console.log('👤 Uploading profile image...');
      
      // Validate file type
      const allowedTypes = ['image/jpeg', 'image/jpg', 'image/png', 'image/webp'];
      if (!allowedTypes.includes(file.type)) {
        throw new Error('Invalid file type. Please upload JPEG, PNG, or WebP images.');
      }
      
      // Validate file size (max 5MB for profiles)
      const maxSize = 5 * 1024 * 1024; // 5MB
      if (file.size > maxSize) {
        throw new Error('File too large. Please upload images smaller than 5MB.');
      }
      
      // Use consistent file ID for profiles (allows easy replacement)
      const fileId = `profile_${userId}`;
      
      // Delete existing profile image if it exists
      try {
        await storage.deleteFile(BUCKETS.profileImages, fileId);
        console.log('🗑️ Deleted existing profile image');
      } catch (deleteError) {
        // File doesn't exist, that's fine
        console.log('ℹ️ No existing profile image to delete');
      }
      
      // Upload new file
      const uploadedFile = await storage.createFile(
        BUCKETS.profileImages,
        fileId,
        file
      );
      
      // Get file URLs
      const fileUrl = getFileView(BUCKETS.profileImages, fileId);
      const thumbnailUrl = getFilePreview(BUCKETS.profileImages, fileId, 150, 150, 85);
      const avatarUrl = getFilePreview(BUCKETS.profileImages, fileId, 64, 64, 80);
      
      console.log('✅ Profile image uploaded:', fileId);
      
      return {
        file: uploadedFile,
        fileId: fileId,
        url: fileUrl,
        thumbnailUrl: thumbnailUrl,
        avatarUrl: avatarUrl,
        filename: file.name,
        size: file.size,
        mimeType: file.type,
        success: true
      };
      
    } catch (error) {
      console.error('❌ Upload profile image failed:', error);
      return {
        success: false,
        error: handleAppwriteError(error)
      };
    }
  }

  // Delete File
  async deleteFile(bucketId, fileId) {
    try {
      console.log('🗑️ Deleting file:', fileId);
      
      await storage.deleteFile(bucketId, fileId);
      
      console.log('✅ File deleted successfully');
      
      return { success: true };
      
    } catch (error) {
      console.error('❌ Delete file failed:', error);
      return {
        success: false,
        error: handleAppwriteError(error)
      };
    }
  }

  // Delete Item Images
  async deleteItemImages(fileIds) {
    try {
      console.log('🗑️ Deleting item images:', fileIds);
      
      const deletePromises = fileIds.map(fileId =>
        this.deleteFile(BUCKETS.itemImages, fileId)
      );
      
      const results = await Promise.all(deletePromises);
      
      const successCount = results.filter(r => r.success).length;
      const errors = results.filter(r => !r.success).map(r => r.error);
      
      console.log(`✅ Deleted ${successCount}/${fileIds.length} images`);
      
      return {
        deleted_count: successCount,
        total_count: fileIds.length,
        errors: errors,
        success: successCount > 0
      };
      
    } catch (error) {
      console.error('❌ Delete item images failed:', error);
      return {
        success: false,
        error: handleAppwriteError(error)
      };
    }
  }

  // Get File Info
  async getFileInfo(bucketId, fileId) {
    try {
      const file = await storage.getFile(bucketId, fileId);
      
      return {
        file,
        success: true
      };
      
    } catch (error) {
      console.error('❌ Get file info failed:', error);
      return {
        success: false,
        error: handleAppwriteError(error)
      };
    }
  }

  // Generate File URLs
  getImageUrls(bucketId, fileId, options = {}) {
    try {
      const {
        thumbnail = { width: 300, height: 300, quality: 85 },
        preview = { width: 800, height: 600, quality: 90 },
        avatar = { width: 64, height: 64, quality: 80 }
      } = options;
      
      return {
        original: getFileView(bucketId, fileId),
        thumbnail: getFilePreview(
          bucketId,
          fileId,
          thumbnail.width,
          thumbnail.height,
          thumbnail.quality
        ),
        preview: getFilePreview(
          bucketId,
          fileId,
          preview.width,
          preview.height,
          preview.quality
        ),
        avatar: getFilePreview(
          bucketId,
          fileId,
          avatar.width,
          avatar.height,
          avatar.quality
        )
      };
      
    } catch (error) {
      console.error('❌ Generate image URLs failed:', error);
      return null;
    }
  }

  // List User Files
  async listUserFiles(userId, bucketId = null) {
    try {
      const bucketsToCheck = bucketId ? [bucketId] : Object.values(BUCKETS);
      const allFiles = [];
      
      for (const bucket of bucketsToCheck) {
        try {
          const files = await storage.listFiles(bucket);
          
          // Filter files that belong to this user (based on naming convention)
          const userFiles = files.files
            .filter(file => {
              // Check if file belongs to user based on metadata or naming
              return (
                file.name.includes(`user_${userId}_`) ||
                file.$id.includes(`profile_${userId}`) ||
                (file.metadata && file.metadata.user_id === userId)
              );
            })
            .map(file => ({
              ...file,
              bucketId: bucket,
              urls: this.getImageUrls(bucket, file.$id)
            }));
          
          allFiles.push(...userFiles);
          
        } catch (bucketError) {
          console.warn(`❌ Failed to list files in bucket ${bucket}:`, bucketError);
        }
      }
      
      return {
        files: allFiles,
        success: true
      };
      
    } catch (error) {
      console.error('❌ List user files failed:', error);
      return {
        success: false,
        error: handleAppwriteError(error)
      };
    }
  }

  // Get Storage Usage Stats
  async getStorageUsage(userId) {
    try {
      const userFilesResult = await this.listUserFiles(userId);
      
      if (!userFilesResult.success) {
        throw new Error('Failed to get user files');
      }
      
      const files = userFilesResult.files;
      
      const usage = {
        total_files: files.length,
        total_size_bytes: files.reduce((sum, file) => sum + (file.sizeOriginal || 0), 0),
        profile_images: files.filter(f => f.bucketId === BUCKETS.profileImages).length,
        item_images: files.filter(f => f.bucketId === BUCKETS.itemImages).length,
        total_size_mb: files.reduce((sum, file) => sum + (file.sizeOriginal || 0), 0) / 1024 / 1024
      };
      
      return {
        usage,
        success: true
      };
      
    } catch (error) {
      console.error('❌ Get storage usage failed:', error);
      return {
        success: false,
        error: handleAppwriteError(error)
      };
    }
  }

  // Batch Upload for Multiple Images
  async uploadMultipleImages(files, userId, listingId = null, bucketType = 'item') {
    try {
      console.log(`📸 Uploading ${files.length} images...`);
      
      // Validate file count
      const maxFiles = bucketType === 'item' ? 10 : 1;
      if (files.length > maxFiles) {
        throw new Error(`Maximum ${maxFiles} files allowed`);
      }
      
      // Upload files in parallel
      const uploadPromises = files.map(file => {
        if (bucketType === 'item') {
          return this.uploadItemImage(file, userId, listingId);
        } else {
          return this.uploadProfileImage(file, userId);
        }
      });
      
      const results = await Promise.all(uploadPromises);
      
      // Separate successful and failed uploads
      const successful = results.filter(r => r.success);
      const failed = results.filter(r => !r.success);
      
      console.log(`✅ Uploaded ${successful.length}/${files.length} images successfully`);
      
      return {
        successful: successful,
        failed: failed,
        total_uploaded: successful.length,
        total_failed: failed.length,
        success: successful.length > 0
      };
      
    } catch (error) {
      console.error('❌ Multiple upload failed:', error);
      return {
        success: false,
        error: handleAppwriteError(error)
      };
    }
  }

  // Image Optimization Helpers
  async compressImage(file, maxWidth = 1200, maxHeight = 1200, quality = 0.8) {
    return new Promise((resolve, reject) => {
      try {
        const canvas = document.createElement('canvas');
        const ctx = canvas.getContext('2d');
        const img = new Image();
        
        img.onload = () => {
          // Calculate new dimensions
          let { width, height } = img;
          
          if (width > maxWidth || height > maxHeight) {
            const ratio = Math.min(maxWidth / width, maxHeight / height);
            width *= ratio;
            height *= ratio;
          }
          
          // Set canvas dimensions
          canvas.width = width;
          canvas.height = height;
          
          // Draw and compress
          ctx.drawImage(img, 0, 0, width, height);
          
          canvas.toBlob(
            (blob) => {
              resolve(new File([blob], file.name, {
                type: 'image/jpeg',
                lastModified: Date.now()
              }));
            },
            'image/jpeg',
            quality
          );
        };
        
        img.onerror = () => reject(new Error('Failed to load image'));
        img.src = URL.createObjectURL(file);
        
      } catch (error) {
        reject(error);
      }
    });
  }

  // Validate Image File
  validateImageFile(file, maxSize = 10 * 1024 * 1024) {
    const errors = [];
    
    // Check file type
    const allowedTypes = ['image/jpeg', 'image/jpg', 'image/png', 'image/webp'];
    if (!allowedTypes.includes(file.type)) {
      errors.push('Invalid file type. Please upload JPEG, PNG, or WebP images.');
    }
    
    // Check file size
    if (file.size > maxSize) {
      const maxSizeMB = maxSize / 1024 / 1024;
      errors.push(`File too large. Please upload images smaller than ${maxSizeMB}MB.`);
    }
    
    // Check if file is actually an image
    if (!file.type.startsWith('image/')) {
      errors.push('Selected file is not an image.');
    }
    
    return {
      isValid: errors.length === 0,
      errors: errors
    };
  }

  // Create Image Thumbnail URL
  createThumbnailUrl(bucketId, fileId, size = 150) {
    return getFilePreview(bucketId, fileId, size, size, 85);
  }

  // Create Responsive Image URLs
  createResponsiveUrls(bucketId, fileId) {
    return {
      small: getFilePreview(bucketId, fileId, 300, 200, 85),
      medium: getFilePreview(bucketId, fileId, 600, 400, 85),
      large: getFilePreview(bucketId, fileId, 1200, 800, 90),
      original: getFileView(bucketId, fileId)
    };
  }
}

// Create and export singleton instance
const appwriteStorage = new AppwriteStorageService();

export default appwriteStorage;